src/app/app.component.ts

```diff
@@ -40,22 +44,50 @@ export class AppComponent implements OnDestroy {
     }
 
     private processSourceChange(newSource: string): void {
-        const result = this._parserService.parse(newSource);
-        if (!result) {
-            return;
-        }
+        const errors = new Set<string>();
+        const result = this._parserService.parse(newSource, errors);
+        this.updateValidation(result, errors);
+
+        if (!result) return;
 
         this._displayService.updateCurrentRun(result);
     }
 
     private processNewSource(newSource: string): void {
-        const result = this._parserService.parse(newSource);
-        if (!result) {
-            return;
-        }
+        const errors = new Set<string>();
+        const result = this._parserService.parse(newSource, errors);
+        this.updateValidation(result, errors);
+
+        if (!result) return;
 
         this._displayService.registerRun(result);
-        this.textareaFc.setValue(newSource, { emitEvent: false });
+        this.updateTextarea(newSource);
+    }
+
+    private updateValidation(
+        run: Run | null,
+        errors: Set<string> = new Set<string>()
+    ): void {
+        this._runHint = [...errors, ...(run ? run.warnings : [])].join('\n');
+
+        if (!run || errors.size > 0) {
+            this.textareaFc.setErrors({ 'invalid run': true });
+            this._isRunValid = Valid.Error;
+        } else if (run.warnings.size > 0) {
+            this._isRunValid = Valid.Warn;
+        } else if (!run.isEmpty()) {
+            this._isRunValid = Valid.Success;
+        } else {
+            this._isRunValid = null;
+        }
+    }
+
+    get isRunValid(): number | null {
+        return this._isRunValid;
+    }
+
+    get runHint(): string {
+        return this._runHint;
     }
 
```

Comment:
Solche TypeScript getter sind immer schlecht in Templates bzw. bei der Verwendung innerhalb von Templates. Für Angular sind getter auch normale Methode-Aufrufe und damit kommt die Change-Detection von Angular leider gar nicht klar.
Methoden werden unfassbar oft von Angular aufgerufen, weil Angular logischerweise nicht erkennt, wenn sich etwas ändert. Das ist anderes, wenn man direkt die Property im Template verwendet.
Das kann man leicht nachvollziehen, wenn man hier in den Getter einfach mal einen `console.log` einbaut

```suggestion
```